network-config-analyzer/GenericTreeScanner.py

```diff
@@ -0,0 +1,95 @@
+#
+# Copyright 2020- IBM Inc. All rights reserved
+# SPDX-License-Identifier: Apache2.0
+#
+
+import os
+import abc
+from enum import Enum
+from dataclasses import dataclass
+from ruamel.yaml import YAML, error
+from sys import stderr
+
+
+@dataclass
+class YamlFile:
+    """
+    A class for holding the retrieved data of a yaml file from git repo
+    """
+    data: object
+    path: str
+
+
+class GenericTreeScanner(metaclass=abc.ABCMeta):
+    """
+    A base class for reading yaml files
+    """
+
+    class ScannerType(Enum):
+        """
+        A class that represents what to be scanned
+        """
+        GitUrl = 0
+        FileSystemPath = 1
+
+    def __init__(self, scanner_type, np_scanner):
+        self.scanner_type = scanner_type
+        self.np_scanner = np_scanner
+
+    @abc.abstractmethod
+    def get_yamls(self):
+        pass
+
+    @staticmethod
+    def is_yaml_file(file_name):
+        """
+        returns if the given file is a yaml file
+        :param str file_name: the name of the file
+        :rtype: bool
+        """
+        extension = os.path.splitext(file_name)[1]
+        return extension in {'.yaml', '.yml', '.json'}
+
+    def _yield_yaml_file(self, path, stream):
+        """
+        yields the yaml file for its data
+        :param str path: the path of the file
+        :param stream: an IO-Text stream or Union of the file contents, depends on the scanner's type
+        """
+        decoded_stream = stream
+        if self.scanner_type == GenericTreeScanner.ScannerType.GitUrl:
+            decoded_stream = stream.decoded_content
+        load_type = None
+        if not self.np_scanner:
+            load_type = "safe"
+        yaml = YAML(typ=load_type)
```

Comment:
```suggestion
        yaml = YAML(typ='safe')
```
We should fix our tests, instead of giving-up safe loading. The fix is rather simple: adding quotes around selectors with '!'.
That is, change
```
      selector: !has(app)
```
into
```
      selector: "!has(app)"
```